[iOS] Mainframe scroll snap jumps to unpredictable offsets when pinch zooming

whsieh opened the pull request:
#### f4b2339d9e5187c28766b443ac6c29c27f59821c
<pre>
[iOS] Mainframe scroll snap jumps to unpredictable offsets when pinch zooming
<a href="https://bugs.webkit.org/show_bug.cgi?id=255848">https://bugs.webkit.org/show_bug.cgi?id=255848</a>
rdar://108008629

Reviewed by Brent Fulgham and Simon Fraser.

When pinch zooming out to less than the initial scale on book.stevejobsarchive.com (when scroll
snapping on the root element is enabled), the scroll position frequently jumps to (seemingly) random
offsets in the page, sometimes after attempting to scroll (with animation) to another wrong scroll
snap point. There are multiple root causes that trigger these symptoms, described below (along with
the changes in this patch that address each one):

(1) We occasionally use the wrong scale to compute the appropriate snap offset in
    `RemoteScrollingCoordinatorProxyIOS`. Currently, `WebPageProxy::displayedContentScale` only
    yields the latest zoom scale we sent via `VisibleContentRectUpdate`, but we scale (and
    counterscale) geometry directly on the `WKScrollView` using this scale. There&apos;s no guarantee
    that `-[UIScrollView zoomScale]` is in sync with this last sent scale, so we can end up
    computing the wrong nearest snap position in both `closestSnapOffsetForMainFrameScrolling` (when
    ending the pinch) and `nearestActiveContentInsetAdjustedSnapOffset` (when we receive the first
    stable-state commit after the bounce animation).

    To fix this, we simply use `UIScrollView`&apos;s `-zoomScale` instead so that the scale is consistent
    with the current value of `-contentOffset`, as well as the retargeted content offset.

(2) Even ignoring (1), the scroll view sometimes fails to retarget to any offset at all after
    pinching out, and instead settles back down to the current offset (as if snapping were
    disabled), before instantly jumping to a nearby snap point. Whether or not this bug happens is
    dependent on the order in which `WKScrollView`&apos;s pan gesture action fires in Ended state —
    relative to the pinch gesture — when ending the pinch. If the scroll view observes the pan
    gesture ending (`-handlePan:`) before the pinch gesture ends (`-handlePinch:`), we&apos;ll run our
    retargeting logic for scroll snapping, but then `UIScrollView` immediately stomps over the
    retargeted position due to the `-zoomScale` clamping back to `-minimumZoomScale` when ending the
    pinch. This ordering is currently non-deterministic, as it depends on hashing order of the
    scroll view&apos;s pinch and pan gesture recognizers, when sending actions to gestures after ending
    the touch.

    To work around this, we take advantage of the fact that `-handlePinch:` is idempotent when the
    pinch gesture has ended (since we&apos;ll clear the `zooming` flag and transition to `zoomBouncing`
    state only once), and directly invoke `-handlePinch:` from within `-handlePan:` here to ensure
    that we&apos;re able to use the retargeting delegate hook.

    There are other workarounds that I considered, such as detecting that this bug is about to
    happen and falling back to an animated scroll instead of retargeting; however, the above is the
    only approach I could find that results in a smooth &quot;glide&quot; animation to the correct snap
    destination (as opposed to zooming back to minimum scale and then scrolling to the final
    location).

(3) There&apos;s logic in `-_updateVisibleContentRects` that&apos;s responsible for scrolling to the active
    snap position when we&apos;re about to send a stable visible content rect update; however, this fires
    way too early in some cases (i.e. right after the pinch ends, in the middle of zooming), causing
    us to jump to the final offset. This is because we send a visible content rect update after the
    `-isZooming` flag is unset (i.e. after the pinch gesture recognizer transitions to Ended state),
    but before the zooming actually begins. Correct this gap in view stability state by additionally
    consulting the `-isZoomBouncing` flag, which is set right after zooming ends when pinching out
    and unset after the zoom animation completes.

(4) After fixing all the above, the behavior is considerably improved, but scrolling still
    occasionally jumps to the final offset in the middle of the zoom bounce animation. This is
    because `resnapAfterLayout()` triggers before the zoom animation is complete, causing a stutter
    in the scroll position as we attempt an instant scroll to the final offset while still zooming.

    Address this by avoiding `resnapAfterLayout()` in the case where the viewport is not in stable
    state; to achieve this, we add some client plumbing through `LocalFrameView` -&gt;
    `WebChromeClient` -&gt; `WebPage` to return whether or not we&apos;re in stable state. Note that we also
    rely on (3) here, otherwise the web process will enter stable state too early.

With the 4 issues above fixed, pinch zooming on book.stevejobsarchive.com no longer causes the
scroll position to visually stutter or jump, and we instead transition in one smooth animation from
the offset in the pinched-out view to the final scroll snap destination, in all cases.

Test: css3/scroll-snap/ios/scroll-snap-mainframe-pinch-out.html

* LayoutTests/css3/scroll-snap/ios/scroll-snap-mainframe-pinch-out-expected.txt: Added.
* LayoutTests/css3/scroll-snap/ios/scroll-snap-mainframe-pinch-out.html: Added.

Add a new test case to exercise this scenario; this test uses scroll snapping on the root element,
and additionally continuously triggers layout on the page by changing the widths of each of the
scroll snap children every frame. The test then programmatically scrolls to a snap point in the
middle of the page, zooms out by pinching, and verifies that after the bounce has ended (i.e. we get
a stable state presentation update), the final scroll offset matches what we started with.

* LayoutTests/css3/scroll-snap/ios/scroll-snap-mainframe-scroll-deceleration-with-obscured-inset.html:

Drive-by fix: clean up this test page by moving the `script` elements in this test under the `head`.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async pinch):

Add a `UIHelper` method to make it easier to synthesize a pinch gesture; this helper method takes
4 points (8 x/y values total) that represent the two starting touch locations, and two ending touch
locations. It then creates and synthesizes a 2-finger gesture stream using these points.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::isInStableState const):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::isInStableState const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::resnapAfterLayout):

See (4) in the above diagnosis.

* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::isInStableState const):
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

Add more forward declarations on `UIScrollView`.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _viewStabilityState:]):

See (3) in the above diagnosis.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::closestSnapOffsetForMainFrameScrolling const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::nearestActiveContentInsetAdjustedSnapOffset const):

See (1) in the above diagnosis.

* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView _sendPinchGestureActionEarlyIfNeeded]):
(-[WKScrollView handlePan:]):

See (2) in the above diagnosis.

* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::isInStableState const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::isInStableState const):

Canonical link: <a href="https://commits.webkit.org/263336@main">https://commits.webkit.org/263336@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8e3e358a614b41cf90c5c4bfffd8508be3d85ce7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4240 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4362 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5711 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4485 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/4234 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4491 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4327 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/4705 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4302 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4487 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3828 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5705 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1982 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3804 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/6183 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3811 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3871 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5402 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4280 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3474 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3790 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3802 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1053 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/7876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4136 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3052,6 +3052,13 @@ bool LocalFrameView::isRubberBandInProgress() const
+{
+    if (auto* page = m_frame->page())
+        return page->chrome().client().isInStableState();
+    return ScrollableArea::isInStableState();
```
Is it reasonable to jump all the way to `ScrollableArea` here? Should we just call into the direct base class in case we need other stability logic in intervening layers?

whsieh replied:
Whoops, good catch! This is just a typo — I intended to just call into the immediate base class.

brentfulgham commented on Source/WebCore/platform/ScrollableArea.cpp:
```diff
@@ -545,7 +545,7 @@ void ScrollableArea::resnapAfterLayout()
 
     auto* scrollAnimator = existingScrollAnimator();
-    if (!scrollAnimator || isScrollSnapInProgress() || isUserScrollInProgress())
+    if (!scrollAnimator || isScrollSnapInProgress() || isUserScrollInProgress() || !isInStableState())
```
This logic makes me feel like `isInStableState()` should encapsulate the in-progress nature of ScrollSnap or UserScroll. I.e., `isInStableState` should probably be false any time a snap or user scroll is in progress.

whsieh replied:
Yeah, I think that makes sense — in particular, the "isUserScrollInProgress" flag should probably have a slightly longer lifetime, to fully account for zoom bouncing on iOS. That said, I do think there's still _some_ value in checking stable state — for instance, if an embedding WebKit client is performing a programmatically-driven animated scroll/zoom on the web view's scroll view.

brentfulgham commented on Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
```diff
@@ -348,13 +349,14 @@
     if (!vertical.isEmpty() && m_currentVerticalSnapPointIndex && *m_currentVerticalSnapPointIndex < vertical.size()) {
-        float potentialSnapPosition = vertical[*m_currentVerticalSnapPointIndex].offset * webPageProxy().displayedContentScale();
+        float potentialSnapPosition = vertical[*m_currentVerticalSnapPointIndex].offset * zoomScale;
         potentialSnapPosition -= topInset;
```
Does the inset not need to be adjusted for zoom?

whsieh replied:
Oh, good question — I mostly just left this logic as-is since it wasn't necessary to fix the broken scroll snapping behavior, but it might be indicative of another bug...I'm double-checking this.

Edit: this existing logic seems correct — we end up snapping to the wrong offset if I try and scale the inset here. I think that makes sense as well, given that the inset itself doesn't get scaled along with the content when pinch zooming.

brentfulgham commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1715,5 +1715,13 @@ bool WebChromeClient::isUsingUISideCompositing() const
+{
+#if PLATFORM(IOS_FAMILY)
+    return m_page.isInStableState();
+#else
```
This is a little surprising -- surely macOS has snapping and rubber-banding that might leave the client in an unstable state. Or is this stability about something that truly doesn't apply for macOS?

whsieh replied:
Ah, yeah — I think ideally, this would be sensibly implemented on macOS as well, but implementing that seemed non-trivial, and in the spirit of keeping the fix targeted I decided to hold off on that part. Definitely warrants a FIXME, though.

brentfulgham reviewed (approved):
What a great job tracking this complex issue down. I had a few questions base on my reading, but I think this seems like a good approach.

Thank you for expanding the testing infrastructure, too. This will likely help us avoid future issues in this space. I wonder if any of it could be published to WPT.

If Simon is around, it would be good to get his thoughts, but I think it looks good so will approve.

whsieh committed:
[iOS] Mainframe scroll snap jumps to unpredictable offsets when pinch zooming
https://bugs.webkit.org/show_bug.cgi?id=255848
rdar://108008629

Reviewed by Brent Fulgham and Simon Fraser.

When pinch zooming out to less than the initial scale on book.stevejobsarchive.com (when scroll
snapping on the root element is enabled), the scroll position frequently jumps to (seemingly) random
offsets in the page, sometimes after attempting to scroll (with animation) to another wrong scroll
snap point. There are multiple root causes that trigger these symptoms, described below (along with
the changes in this patch that address each one):

(1) We occasionally use the wrong scale to compute the appropriate snap offset in
    `RemoteScrollingCoordinatorProxyIOS`. Currently, `WebPageProxy::displayedContentScale` only
    yields the latest zoom scale we sent via `VisibleContentRectUpdate`, but we scale (and
    counterscale) geometry directly on the `WKScrollView` using this scale. There's no guarantee
    that `-[UIScrollView zoomScale]` is in sync with this last sent scale, so we can end up
    computing the wrong nearest snap position in both `closestSnapOffsetForMainFrameScrolling` (when
    ending the pinch) and `nearestActiveContentInsetAdjustedSnapOffset` (when we receive the first
    stable-state commit after the bounce animation).

    To fix this, we simply use `UIScrollView`'s `-zoomScale` instead so that the scale is consistent
    with the current value of `-contentOffset`, as well as the retargeted content offset.

(2) Even ignoring (1), the scroll view sometimes fails to retarget to any offset at all after
    pinching out, and instead settles back down to the current offset (as if snapping were
    disabled), before instantly jumping to a nearby snap point. Whether or not this bug happens is
    dependent on the order in which `WKScrollView`'s pan gesture action fires in Ended state —
    relative to the pinch gesture — when ending the pinch. If the scroll view observes the pan
    gesture ending (`-handlePan:`) before the pinch gesture ends (`-handlePinch:`), we'll run our
    retargeting logic for scroll snapping, but then `UIScrollView` immediately stomps over the
    retargeted position due to the `-zoomScale` clamping back to `-minimumZoomScale` when ending the
    pinch. This ordering is currently non-deterministic, as it depends on hashing order of the
    scroll view's pinch and pan gesture recognizers, when sending actions to gestures after ending
    the touch.

    To work around this, we take advantage of the fact that `-handlePinch:` is idempotent when the
    pinch gesture has ended (since we'll clear the `zooming` flag and transition to `zoomBouncing`
    state only once), and directly invoke `-handlePinch:` from within `-handlePan:` here to ensure
    that we're able to use the retargeting delegate hook.

    There are other workarounds that I considered, such as detecting that this bug is about to
    happen and falling back to an animated scroll instead of retargeting; however, the above is the
    only approach I could find that results in a smooth "glide" animation to the correct snap
    destination (as opposed to zooming back to minimum scale and then scrolling to the final
    location).

(3) There's logic in `-_updateVisibleContentRects` that's responsible for scrolling to the active
    snap position when we're about to send a stable visible content rect update; however, this fires
    way too early in some cases (i.e. right after the pinch ends, in the middle of zooming), causing
    us to jump to the final offset. This is because we send a visible content rect update after the
    `-isZooming` flag is unset (i.e. after the pinch gesture recognizer transitions to Ended state),
    but before the zooming actually begins. Correct this gap in view stability state by additionally
    consulting the `-isZoomBouncing` flag, which is set right after zooming ends when pinching out
    and unset after the zoom animation completes.

(4) After fixing all the above, the behavior is considerably improved, but scrolling still
    occasionally jumps to the final offset in the middle of the zoom bounce animation. This is
    because `resnapAfterLayout()` triggers before the zoom animation is complete, causing a stutter
    in the scroll position as we attempt an instant scroll to the final offset while still zooming.

    Address this by avoiding `resnapAfterLayout()` in the case where the viewport is not in stable
    state; to achieve this, we add some client plumbing through `LocalFrameView` ->
    `WebChromeClient` -> `WebPage` to return whether or not we're in stable state. Note that we also
    rely on (3) here, otherwise the web process will enter stable state too early.

With the 4 issues above fixed, pinch zooming on book.stevejobsarchive.com no longer causes the
scroll position to visually stutter or jump, and we instead transition in one smooth animation from
the offset in the pinched-out view to the final scroll snap destination, in all cases.

Test: css3/scroll-snap/ios/scroll-snap-mainframe-pinch-out.html

* LayoutTests/css3/scroll-snap/ios/scroll-snap-mainframe-pinch-out-expected.txt: Added.
* LayoutTests/css3/scroll-snap/ios/scroll-snap-mainframe-pinch-out.html: Added.

Add a new test case to exercise this scenario; this test uses scroll snapping on the root element,
and additionally continuously triggers layout on the page by changing the widths of each of the
scroll snap children every frame. The test then programmatically scrolls to a snap point in the
middle of the page, zooms out by pinching, and verifies that after the bounce has ended (i.e. we get
a stable state presentation update), the final scroll offset matches what we started with.

* LayoutTests/css3/scroll-snap/ios/scroll-snap-mainframe-scroll-deceleration-with-obscured-inset.html:

Drive-by fix: clean up this test page by moving the `script` elements in this test under the `head`.

* LayoutTests/resources/ui-helper.js:
(window.UIHelper.async pinch):

Add a `UIHelper` method to make it easier to synthesize a pinch gesture; this helper method takes
4 points (8 x/y values total) that represent the two starting touch locations, and two ending touch
locations. It then creates and synthesizes a 2-finger gesture stream using these points.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::isInStableState const):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::isInStableState const):
* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/ScrollableArea.cpp:
(WebCore::ScrollableArea::resnapAfterLayout):

See (4) in the above diagnosis.

* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::isInStableState const):
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

Add more forward declarations on `UIScrollView`.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _viewStabilityState:]):

See (3) in the above diagnosis.

* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::closestSnapOffsetForMainFrameScrolling const):
(WebKit::RemoteScrollingCoordinatorProxyIOS::nearestActiveContentInsetAdjustedSnapOffset const):

See (1) in the above diagnosis.

* Source/WebKit/UIProcess/ios/WKScrollView.mm:
(-[WKScrollView _sendPinchGestureActionEarlyIfNeeded]):
(-[WKScrollView handlePan:]):

See (2) in the above diagnosis.

* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::isInStableState const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::isInStableState const):

Canonical link: https://commits.webkit.org/263336@main
